osu.Server.Spectator/Entities/ConnectionState.cs

```diff
@@ -42,5 +58,37 @@ public ConnectionState(HubLifetimeContext context)
         /// <param name="context">The hub context to retrieve information from.</param>
         public void RegisterConnectionId(HubLifetimeContext context)
             => ConnectionIds[context.Hub.GetType()] = context.Context.ConnectionId;
+
+        public bool IsConnectionFromSameClient(HubLifetimeContext context)
+        {
+            if (tryGetClientSessionID(context, out var clientSessionId))
+                return ClientSessionId == clientSessionId;
+
+            // Legacy pathway using JTI claim left for compatibility with older clients – can be removed 2024-08-18
+            return TokenId == context.Context.GetTokenId();
+        }
+
+        public bool IsInvocationPermitted(HubInvocationContext context)
+        {
+            bool hubRegistered = ConnectionIds.TryGetValue(context.Hub.GetType(), out string? registeredConnectionId);
+            bool connectionIdMatches = registeredConnectionId == context.Context.ConnectionId;
+
+            return hubRegistered && connectionIdMatches;
+        }
+
+        public bool CanCleanUpConnection(HubLifetimeContext context)
+        {
+            bool hubRegistered = ConnectionIds.TryGetValue(context.Hub.GetType(), out string? registeredConnectionId);
+            bool connectionIdMatches = registeredConnectionId == context.Context.ConnectionId;
+
+            return hubRegistered && connectionIdMatches;
```

Comment:
Is there some special reason for having these two methods exist rather than a single method? They are doing the exact same thing now.

Reply:
`HubLifetimeContext` and `HubInvocationContext` do not have a common superclass despite the methods in question indeed being mirror images of each other.

Reply:
right, missed that. having the method name the same might help in that case, if you can think of a name which works for both.

Reply:
nevermind, what you PR'd works fine.